cloudwatch_metrics_collector/test/metrics_collector_test.cpp

```diff
@@ -0,0 +1,238 @@
+  std::shared_ptr<MetricServiceFactoryMock> mock_metric_service_factory_;
+  MetricsCollector metrics_collector_;
+};
+
+constexpr char MetricsCollectorFixture::kTestMetricsNamespace[];
+
+TEST_F(MetricsCollectorFixture, RecordRosMonitoringMessages)
+{
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_real_distribution<> dist(0.0, 100.0);
+
+  std::array<ros_monitoring_msgs::msg::MetricData, 2> metric_data = {{
+    BasicMonitoringData(),
+    BasicMonitoringData()
+  }};
+  for (auto & data : metric_data) {
+    data.value = dist(gen);
+  }
+
+  auto metric_list = std::make_unique<ros_monitoring_msgs::msg::MetricList>();
+  for (const auto & data : metric_data) {
+    metric_list->metrics.push_back(data);
+  }
+
+  size_t num_batched = metrics_collector_.RecordMetrics(std::move(metric_list));
+  ASSERT_EQ(metric_data.size(), num_batched);
+
+  const auto & metric_objs = mock_metric_service_->GetMetricObjects();
+  ASSERT_EQ(metric_data.size(), metric_objs.size());
+
+  for (size_t i = 0; i < metric_data.size(); ++i) {
+    EXPECT_EQ(metric_data[i].metric_name, metric_objs[i].metric_name);
+    EXPECT_EQ(metric_data[i].unit, metric_objs[i].unit);
+    if (std::isnan(metric_data[i].value)) {
+      EXPECT_TRUE(std::isnan(metric_objs[i].value));
+    } else {
+      EXPECT_DOUBLE_EQ(metric_data[i].value, metric_objs[i].value);
+    }
+    EXPECT_EQ(MetricsCollector::GetMetricDataEpochMillis(metric_data[i].time_stamp), metric_objs[i].timestamp);
+
+    std::map<std::string, std::string> dimensions;
+    for (const auto & dimension : metric_data[i].dimensions) {
+      dimensions[dimension.name] = dimension.value;
+    }
+    EXPECT_EQ(dimensions, metric_objs[i].dimensions);
+  }
+}
+
+TEST_F(MetricsCollectorFixture, RecordMetricsMessages)
+{
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::uniform_real_distribution<> dist(0.0, 100.0);
+
+  std::array<metrics_statistics_msgs::msg::MetricsMessage, 2> metric_data = {{
+    BasicMetricsData(),
+    BasicMetricsData()
+  }};
+  for (auto & metric : metric_data) {
+    for (auto & statistic : metric.statistics) {
+      statistic.data = dist(gen);
+    }
+  }
+
+  for (const auto & data : metric_data) {
+    auto msg = std::make_unique<metrics_statistics_msgs::msg::MetricsMessage>(data);
+    int num_batched = metrics_collector_.RecordMetrics(std::move(msg));
+    ASSERT_EQ(2, num_batched);
+  }
+
+  const auto & metric_objs = mock_metric_service_->GetMetricObjects();
+  ASSERT_EQ(2 * metric_data.size(), metric_objs.size());
+
+  for (size_t i = 0; i < metric_data.size(); ++i) {
+    EXPECT_EQ(metric_data[i].measurement_source_name, metric_objs[2*i].dimensions.at("Measurement Source"));
+    EXPECT_EQ(metric_data[i].measurement_source_name, metric_objs[2*i+1].dimensions.at("Measurement Source"));
+
+    EXPECT_EQ(metric_data[i].metrics_source, metric_objs[2*i].metric_name);
+    EXPECT_EQ(metric_data[i].metrics_source+"_stddev", metric_objs[2*i+1].metric_name);
```

Comment:
Ditto as above. Dislike this math here -- why do we need all this ?
Just test if what you set is what you get.

Reply:
`MetricsCollector::RecordMetrics()` does this precise manipulation as part of mapping the standard deviation data to `MetricObject`s, and this is for verifying that manipulation is being done.

Reply:
> `MetricsCollector::RecordMetrics()` does this precise manipulation as part of mapping the standard deviation data to `MetricObject`s, and this is for verifying that manipulation is being done.

Maybe it does. But this can be a very simple test (or collection of tests). 
Test 1: TestMetricsDataWithStddev  -> should produce 2 metrics objects. 
Test 2: TestMetricsDataWithoutStdDev --> should produce only one metric object

Then its reduces to just checking for the generated metrics objects are the values accurate.

Reply:
Done.